File: Java/DesignPatterns/src/main/java/designPatterns/behavioral/chain/ConcreteHandler.java
package designPatterns.behavioral.chain;

public class ConcreteHandler {
  static class A extends Handler {
    @Override
    public String handleRequest(String request) {
      if (request.equals("one")) {
        return "ConcreteHandlerA";
      } else if (next != null) {
        return next.handleRequest(request);
      }
      return null;
    }
  }

  static class B extends Handler {
    @Override
    public String handleRequest(String request) {
      if (request.equals("two")) {
        return "ConcreteHandlerB";
      } else if (next != null) {
        return next.handleRequest(request);
      }
      return null;
    }
  }
}

File: Kotlin/DesignPatterns/src/main/kotlin/behavioral/chain/ConcreteHandler.kt
package behavioral.chain

sealed class ConcreteHandler {
  class A : Handler() {
    override fun handleRequest(request: String): String? {
      if (request.equals("one")) {
        return "ConcreteHandlerA";
      } else if (next != null) {
        return next!!.handleRequest(request)
      }
      return null;
    }
  }

  class B : Handler() {
    override fun handleRequest(request: String): String? {
      if (request.equals("two")) {
        return "ConcreteHandlerB";
      } else if (next != null) {
        return next!!.handleRequest(request);
      }
      return null;
    }
  }
}
